lib/host/wasi/inode-linux.cpp

```diff
@@ -156,11 +156,33 @@ void DirHolder::reset() noexcept {
   }
 }
 
-INode INode::stdIn() noexcept { return INode(STDIN_FILENO); }
+namespace {
+WasiExpect<INode> createStdNode(int32_t Fd, const char *streamName) {
+  if (Fd < 0) {
+    spdlog::error("    Invalid {} file descriptor: {}"sv, streamName, Fd);
+    return WasiUnexpect(__WASI_ERRNO_BADF);
+  }
+
+  if (fcntl(Fd, F_GETFD) == -1) {
+    spdlog::error("    Invalid {} file descriptor: {}"sv, streamName, Fd);
+    return WasiUnexpect(__WASI_ERRNO_BADF);
+  }
 
-INode INode::stdOut() noexcept { return INode(STDOUT_FILENO); }
+  return INode(Fd, false, false);
+}
+} // namespace
 
-INode INode::stdErr() noexcept { return INode(STDERR_FILENO); }
+WasiExpect<INode> INode::stdIn(int32_t Fd) {
+  return createStdNode(Fd, "stdin");
+}
+
+WasiExpect<INode> INode::stdOut(int32_t Fd) {
+  return createStdNode(Fd, "stdout");
+}
+
+WasiExpect<INode> INode::stdErr(int32_t Fd) {
+  return createStdNode(Fd, "stderr");
+}
```

Comment:
You can keep the original ones, and have a new function called `WasiExpect<INode> INode::fromFd(int32_t Fd)` for all three functions.